src/main/java/org/xmldb/api/security/Permissions.java

```diff
@@ -0,0 +1,357 @@
+            break;
+          default:
+            throw new IllegalArgumentException("Unrecognised mode char '" + c + "'");
+        }
+      }
+      perms.clear();
+    }
+  }
+
+  private static void parseOperation(final String[] whoPerm,
+      final EnumMap<PermType, Boolean> perms) {
+    for (final char c : whoPerm[1].toCharArray()) {
+      switch (c) {
+        case READ_CHAR:
+          perms.put(PermType.READ, TRUE);
+          break;
+        case WRITE_CHAR:
+          perms.put(PermType.WRITE, TRUE);
+          break;
+        case EXECUTE_CHAR:
+          perms.put(PermType.EXECUTE, TRUE);
+          break;
+        case SETID_CHAR:
+          perms.put(PermType.SETID, TRUE);
+          break;
+        case STICKY_CHAR:
+          perms.put(PermType.STICKY, TRUE);
+          break;
+        default:
+          throw new IllegalArgumentException("Unrecognised mode char '" + c + "'");
+      }
+    }
+  }
+
+  private static void setPermissions(EnumSet<Permission> permissions,
+      EnumMap<PermType, Boolean> perms, Permission read, Permission write, Permission execute) {
+    if (perms.containsKey(PermType.READ)) {
+      permissions.add(read);
+    }
+    if (perms.containsKey(PermType.WRITE)) {
+      permissions.add(write);
+    }
+    if (perms.containsKey(PermType.EXECUTE)) {
+      permissions.add(execute);
+    }
+  }
+
+  private static void setSimpleSymbolicMode(final EnumSet<Permission> permissions,
+      final String simpleModeStr) {
+    final char[] modeArray = simpleModeStr.toCharArray();
+    for (int i = 0; i < modeArray.length; i++) {
+      final char c = modeArray[i];
+      switch (c) {
+        case READ_CHAR:
+          setPermission(permissions, i, OWNER_READ, GROUP_READ, OTHERS_READ);
+          break;
+        case WRITE_CHAR:
+          setPermission(permissions, i - 1, OWNER_WRITE, GROUP_WRITE, OTHERS_WRITE);
+          break;
+        case EXECUTE_CHAR:
+          setPermission(permissions, i - 2, OWNER_EXECUTE, GROUP_EXECUTE, OTHERS_EXECUTE);
+          break;
+        case SETID_CHAR_NO_EXEC:
+          if (i < 3) {
+            permissions.add(SET_UID);
+          } else {
+            permissions.add(SET_GID);
+          }
+          break;
+        case SETID_CHAR:
+          if (i < 3) {
+            permissions.add(OWNER_EXECUTE);
+            permissions.add(SET_UID);
+          } else {
+            permissions.add(GROUP_EXECUTE);
+            permissions.add(SET_GID);
+          }
+          break;
+        case STICKY_CHAR_NO_EXEC:
+          permissions.add(STICKY_BIT);
+          break;
+        case STICKY_CHAR:
+          permissions.add(OTHERS_EXECUTE);
+          permissions.add(STICKY_BIT);
+          break;
+        case UNSET_CHAR:
+          break;
+        default:
+          throw new IllegalArgumentException("Unrecognised mode char '" + c + "'");
+      }
+    }
+  }
+
+  private static void setPermission(EnumSet<Permission> permissions, int index, Permission owner,
```

Comment:
I don't understand this function - it appears to take 3 permission args, but will only ever use 1 of them. Why not just have 1 permission arg?

Reply:
The `permissions` argument contains the set of all parsed permissions, the `index` defines the which of the three passed in permissions (`owner`, `group` or 'others`) representing the correct _read_, _write_ or _execute_ permission.

Reply:
I am, not sure I understand... The following code would do the same wouldn't it?

```java
private static void setPermission(EnumSet<Permission> permissions, int index, Permission permission) {
      Permission group, Permission others) {
    switch (index) {
      case 0:
        permissions.add(permission);
        break;
      case 3:
        permissions.add(permission);
        break;
      case 6:
        permissions.add(permission);
        break;
      default:
    }
  }
}
```